enhancements/kubelet/virtualization-higher-workload-density.md

```diff
@@ -0,0 +1,544 @@
+   documentation
+
+   a. The cluster admin is adding the `failOnSwap=false` flag to the
+      kubelet configuration via a `KubeletConfig` CR, in order to ensure
+      that the kubelets will start once swap has been rolled out.
+   a. The cluster admin is calculating the amount of swap space to
+      provision based on the amount of physical ram and overcommitment
+      ratio
+   b. The cluster admin is creating a `MachineConfig` for provisioning
+      swap on worker nodes
+   c. The cluster admin is deploying the [WASP Agent] DaemonSet
+
+4. The cluster admin is configuring OpenShift Virtualization for higher
+   workload density via
+
+   a. the OpenShift Virtualization Console "Settings" page 
+   b. [or `HCO` API](https://github.com/kubevirt/hyperconverged-cluster-operator/blob/main/docs/cluster-configuration.md#configure-higher-workload-density)
+
+The cluster is now set up for higher workload density.
+
+In phase 3, deploying the WASP agent will not be needed.
+
+#### Workflow: Leveraging higher workload density
+
+1. The VM Owner is creating a regular virtual machine and is launching it. The VM owner must not specify memory requests in the VM spec, but only the guest memory size.
+
+### API Extensions
+
+This proposal does not require any Kubernetes, OpenShift, or OpenShift
+Virtualization API changes.
+
+### Topology Considerations
+
+#### Hypershift / Hosted Control Planes
+
+The `MachineConfig` based swap provisioning will not work, as HCP does
+not provide the `MachineConfig` APIs.
+
+#### Standalone Clusters
+
+Standalone, regular and compact clusters are the primary use-cases for
+swap.
+
+#### Single-node Deployments or MicroShift
+
+Single-node and MicroShift deployments are out of scope of this
+proposal.
+
+### Implementation Details/Notes/Constraints
+
+#### WASP Agent
+
+At it's core the [wasp agent] is a `DaemonSet` delivering an [OCI Hook]
+which is used to turn on swap for selected workloads.
+
+Overall the agent is intended to align - and specifically not
+conflict - with the upstream Kubernetes SWAP design and behavior in
+order to simplify a transition.
+
+##### Design
+
+The design is driven by the following guiding principles:
+
+* System services are more important than workloads. Because workload
+  health depends on the health of system services.
+* Try to stay aligned to upstream Kubernetes swap behavior in order
+  to ease the transition to Kubernetes SWAP once available
+
+###### Enabling swap
+
+An OCI Hook to enable swap by setting the containers cgroup
+`memory.swap.max=max`.
+
+* **Tech Preview**
+  * Uses `UnlimitedSwap`.
+* **General Availability**
+  * Limited to burstable QoS class pods.
+  * Uses `LimitedSwap`.
+  * Limited to non-high-priority pods.
+
+For more info, refer to the upstream documentation on how to calculate
+[limited swap](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2400-node-swap#steps-to-calculate-swap-limit).
+
```

Comment:
I would benefit from seeing a diagram (or a link to a diagram) showing the execution order of the CRI, OCI hook and Wasp update.

I would also state clearly that we have a race condition here: container binary may allocate memory before WASP sets a limited value.